src/inc/til/rapidhash.h

```diff
@@ -0,0 +1,384 @@
+    uint64_t a, b;
+    b = __umulh(*A, *B);
+    a = *A * *B;
+    *A ^= a;
+    *B ^= b;
+#else
+    uint64_t c;
+    c = __umulh(*A, *B);
+    *A = *A * *B;
+    *B = c;
+#endif
+#endif
+#else
+    uint64_t ha = *A >> 32, hb = *B >> 32, la = (uint32_t)*A, lb = (uint32_t)*B, hi, lo;
+    uint64_t rh = ha * hb, rm0 = ha * lb, rm1 = hb * la, rl = la * lb, t = rl + (rm0 << 32), c = t < rl;
+    lo = t + (rm1 << 32);
+    c += lo < t;
+    hi = rh + (rm0 >> 32) + (rm1 >> 32) + c;
+#ifdef RAPIDHASH_PROTECTED
+    *A ^= lo;
+    *B ^= hi;
+#else
+    *A = lo;
+    *B = hi;
+#endif
+#endif
+}
+
+/*
+ *  Multiply and xor mix function.
+ *  
+ *  @param A  64-bit number.
+ *  @param B  64-bit number.
+ *
+ *  Calculates 128-bit C = A * B.
+ *  Returns 64-bit xor between high and low 64 bits of C.
+ */
+static inline uint64_t rapid_mix(uint64_t A, uint64_t B) RAPIDHASH_NOEXCEPT
+{
+    rapid_mum(&A, &B);
+    return A ^ B;
+}
+
+/*
+ *  Read functions.
+ */
+#ifdef RAPIDHASH_LITTLE_ENDIAN
+static inline uint64_t rapid_read64(const uint8_t* p) RAPIDHASH_NOEXCEPT
+{
+    uint64_t v;
+    memcpy(&v, p, 8);
+    return v;
+}
+static inline uint64_t rapid_read32(const uint8_t* p) RAPIDHASH_NOEXCEPT
+{
+    uint32_t v;
+    memcpy(&v, p, 4);
+    return v;
+}
+#elif defined(__GNUC__) || defined(__INTEL_COMPILER) || defined(__clang__)
+static inline uint64_t rapid_read64(const uint8_t* p)
+{
+    uint64_t v;
+    memcpy(&v, p, 8);
+    return __builtin_bswap64(v);
+}
+static inline uint64_t rapid_read32(const uint8_t* p)
+{
+    uint32_t v;
+    memcpy(&v, p, 4);
+    return __builtin_bswap32(v);
+}
+#elif defined(_MSC_VER)
+static inline uint64_t rapid_read64(const uint8_t* p)
+{
+    uint64_t v;
+    memcpy(&v, p, 8);
+    return _byteswap_uint64(v);
+}
+static inline uint64_t rapid_read32(const uint8_t* p)
+{
+    uint32_t v;
+    memcpy(&v, p, 4);
+    return _byteswap_ulong(v);
+}
+#else
+static inline uint64_t rapid_read64(const uint8_t* p)
+{
+    uint64_t v;
+    memcpy(&v, p, 8);
+    return (((v >> 56) & 0xff) | ((v >> 40) & 0xff00) | ((v >> 24) & 0xff0000) | ((v >> 8) & 0xff000000) | ((v << 8) & 0xff00000000) | ((v << 24) & 0xff0000000000) | ((v << 40) & 0xff000000000000) | ((v << 56) & 0xff00000000000000));
+}
+static inline uint64_t rapid_read32(const uint8_t* p)
+{
+    uint32_t v;
+    memcpy(&v, p, 4);
+    return (((v >> 24) & 0xff) | ((v >> 8) & 0xff00) | ((v << 8) & 0xff0000) | ((v << 24) & 0xff000000));
+}
+#endif
+
+/*
+ *  Reads and combines 3 bytes of input.
+ *
+ *  @param p  Buffer to read from.
+ *  @param k  Length of @p, in bytes.
+ *
+ *  Always reads and combines 3 bytes from memory.
+ *  Guarantees to read each buffer position at least once.
+ *  
+ *  Returns a 64-bit value containing all three bytes read. 
+ */
+static inline uint64_t rapid_readSmall(const uint8_t* p, size_t k) RAPIDHASH_NOEXCEPT { return ((static_cast<uint64_t>(p[0])) << 56) | ((static_cast<uint64_t>(p[k >> 1])) << 32) | p[k - 1]; }

```

Comment:
But if we do end up vendoring the `rapidhash.h` header I think it'd be best to keep it entirely unmodified. If the "AuditMode" compilation complains or any other part of the CI, I'd rather disable the checks.

Reply:
Talked offline, decided to move header file to oss folder